### PR TITLE
Indoor directions route for disabled

### PR DESCRIPTION
**Pull request by samjamhen**
#### Feature
- useState added in the code: isDisabled
- if it is set to true, only considers paths that lead to elevators
- Also refactored graph and coordinate files to split them so that it is easier to change them individually in the future 
        - will wait before deleting old file

#### Tests
- automated tests
- manual tests

#### How to test PR:
Hall: H919 to H827
VL: VL109 to VL209

#### Expected output
(Please note that for whatever reason, the nodes between the graphs are not properly linked to neighbors and is why the routes sometimes go in diagonal as can be seen on the screenshots. This is a known limitation that will be fixed in another PR to come. It will just have to be a PR fixing the problems with our graphs.)
For Hall:
Screenshot 2025-03-17 at 6 37 16 PM

Screenshot 2025-03-17 at 6 37 22 PM

For VL:
Screenshot 2025-03-17 at 6 36 40 PM
Screenshot 2025-03-17 at 6 36 46 PM

### CODE REVIEW

File: myApp/app/(tabs)/map.jsx [hunk 1]
```diff
@@ -70,6 +70,9 @@ export default function MapView() {
   const [activeCampus, setActiveCampus] = useState(type);
   const mapRef = useRef(null);
 
+  // Disabled on or off
+  const [isDisabled, setIsDisabled] = useState(false);
```
Comment (samjamhen) [inline myApp/app/(tabs)/map.jsx:74]: To test the PR you will have to set this to true as we do not have the frontend yet

**Comment by samjamhen** [comment]
Screenshot 2025-03-17 at 11 41 49 PM
Screenshot 2025-03-17 at 11 41 59 PM

fixed some paths in new commit

File: myApp/app/__tests__/IndoorMapTests/Algorithm.test.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { dijkstra } from "../../components/IndoorMap/ShortestPath";
+
+describe("Dijkstra's Algorithm with Disabled Option", () => {
```
Comment (Carmoucha) [inline myApp/app/__tests__/IndoorMapTests/Algorithm.test.js:3]: The test cases are well-structured but I think it can be improved by adding edge cases to cover scenarios that may cause unexpected behaviour. 

Some examples:
- Start or end node is inaccessible – If navigation starts from a node that only connects through stairs or escalators, the function should immediately return null when isDisabled = true, as there’s no valid way to proceed.
- No valid accessible path exists – If all possible routes require stairs or escalators, the function should recognize this and return null. This prevents cases where a user might be led to an inaccessible location.

These type of tests could improve coverage.

**Review by Carmoucha — COMMENTED** [review]
Great work on implementing the `isDisabled` functionality! The algorithm effectively ensures accessibility by avoiding stairs and escalators while still finding the best possible routes. The logic is well-structured. There are just a few small improvements that could enhance readability and maintainability, such as extracting the node filtering logic into a helper function and adding a couple of edge cases to the test suite. These are minor refinements, but overall, this is good work.

File: myApp/app/components/IndoorMap/ShortestPath.js [hunk 3]
```diff
@@ -1,38 +1,67 @@
...
+export function dijkstra(graph, startNode, endNode, isDisabled) {
+  // this keeps of the distance from the start node to every other node 
   let distances = {};
+  // stores the previous node leading to the shortest path
   let previousNodes = {};
+  // keeps track of nodes that have not been visited yet. initialised with all nodes in the graph
   let unvisitedNodes = new Set(Object.keys(graph));
 
+  // check to see if user requires disabled friendly directions (no stairs, no escalators, only elevators)
+  const shouldSkipNeighbour = (disabled, neighbouringNode) => {
```
Comment (samjamhen) [inline myApp/app/components/IndoorMap/ShortestPath.js:10]: Here is the extracted function to answer carmen's comment

File: myApp/app/__tests__/IndoorMapTests/Algorithm.test.js [hunk 4]
```diff
@@ -29,4 +35,19 @@ describe("Dijkstra's Algorithm with Disabled Option", () => {
...
+  it("returns null if start node is a stair", () => {
+    const result = dijkstra(graph, "stair-Z", "Z", true);
+    expect(result).toBeNull();
+  });
+
+  it ("returns null if the end node is an escalator", () => {
+    const result = dijkstra(graph, "Y", "escalator-Y", true);
+    expect(result).toBeNull();
+  })
+
```
Comment (samjamhen) [inline myApp/app/__tests__/IndoorMapTests/Algorithm.test.js:52]: Unit tests added here to satisfy recommendation from carmen

**Comment by Carmoucha** [comment]
Great work on these changes! Extracting the filtering logic into shouldSkipNeighbour() improves readability, and the additional test cases make the test coverage more complete. The edge cases for inaccessible nodes ensure that the algorithm correctly handles restricted paths when isDisabled is true. Everything looks good now!
